### PR TITLE
Bundle packages

### PR DESCRIPTION
**Pull request by paescuj**
Enable bundling of the following packages:
- `@directus/api`
- `@directus/composables`
- `@directus/constants`
- `@directus/extensions-sdk`
- `@directus/pressure`
- `@directus/schema`
- `@directus/storage-driver-azure`
- `@directus/storage-driver-cloudinary`
- `@directus/storage-driver-gcs`
- `@directus/storage-driver-local`
- `@directus/storage-driver-s3`
- `@directus/storage`
- `@directus/update-check`
- `@directus/utils`

### Notes

- **`@directus/api`**
  - ⚠️ This is a breaking change since exposed  entry points are reduced to  `index.js` (`createApp`) and `cli/run.js` only!
  - Bundling requires two workarounds:
    - In order to be able to get the `yaml` & `liquid` files from the filesystem, the path must be computed dynamically depending on the context (local dev vs. bundled version). For this the util [`findContextRoot`](https://github.com/directus/directus/pull/19089/files#diff-e47968a1a045eda0ed8db3525151bb8a5cb7962067c87d8dba2f355af7e24fb4) has been introduced.
    - Previously dynamically loaded migration and operations scripts are now statically loaded and thus included in the bundle.
    - We should be able to get rid of both workarounds using dynamic imports as soon as https://github.com/evanw/esbuild/pull/2508 has been merged & released. But since we don't know when that will happen, I think it's worth using those workarounds for now.
  - Aligned imports of Node.js Built-ins (e.g. `node:path`) which should lead to somewhat cleaner bundle results.
- **`@directus/types`**
  - Bundling is not possible (and also not really necessary) since values used for `z.infer`(zod) aren't detected by `tsup` and wouldn't be included in the bundle.
  - ⚠️ This pull request however updates the `@directus/types` package to only publish types (`d.ts`) files and no longer the empty `js` files. This is a breaking change because types should now be imported via [`import type`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export) syntax! 
    - Most of the file changes in this pull request come from this change ([befa7c4](https://github.com/directus/directus/pull/19089/commits/befa7c40e7141652ea5f1c6fee10c4eee6176fe4) - update `@directus/app` to use `import type` syntax for `@directus/types`).
  - ~Moving dependencies referenced in this package from `devDependencies` to `dependencies`, see https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#dependencies.~ Adressed in #19296.
- **`@directus/extensions-sdk`**
  - Similar workaround as in `@directus/api` to get correct path to `templates`.
- **`@directus/release-notes-generator`**
  - We don't need types, but if we bundle without types we loose type checking. Therefore and also since bundling isn't really necessary, it will be left as it is.

Closes ENG-1089

### CODE REVIEW

**Review by br41nslug — COMMENTED** [review]
For `@directus/types` you mention the following:
> Bundling is not possible (and also not really necessary) since values used for z.infer(zod) aren't detected by tsup and wouldn't be included in the bundle.

And im curious why this is not an issue for the `@directus/api` package, or is that only applicable to types being exported for external use 🤔 https://github.com/directus/directus/blob/main/api/src/websocket/messages.ts#L57C1-L57C83

Besides that it works as expected in my limited testing.

**Comment by paescuj** [comment]
> And im curious why this is not an issue for the @directus/api package, or is that only applicable to types being exported for external use 🤔 https://github.com/directus/directus/blob/main/api/src/websocket/messages.ts#L57C1-L57C83

Good question! Yeah, it's only an issue when types are being exported 👍

File: api/src/app.ts [hunk 1]
```diff
@@ -2,11 +2,11 @@ import { handlePressure } from '@directus/pressure';
 import cookieParser from 'cookie-parser';
 import type { Request, RequestHandler, Response } from 'express';
 import express from 'express';
-import type { ServerResponse } from 'http';
+import type { ServerResponse } from 'node:http';
 import { merge } from 'lodash-es';
 import { readFile } from 'node:fs/promises';
 import { createRequire } from 'node:module';
-import path from 'path';
+import path from 'node:path';
```
Comment (jaads) [inline api/src/app.ts:9]: just out of curiosity: was the `node` prefix necessary for the new bundling process or just to be more expressive?

Comment (paescuj) [reply api/src/app.ts:9]: 😃 


To expand on that:
* Nowadays, using the `node:` prefix to import Node.js built-in modules is the recommended way for various reasons
* The idea here is to preemptively ensure that  the bundle won't contain multiple import statement for the same module (`path` and `node:path`)

Comment (jaads) [reply api/src/app.ts:9]: Uups 😄👍 alright, thanks!!

**Comment by jaads** [comment]
Shall add a reminder somewhere to keep the issue  https://github.com/evanw/esbuild/pull/2508 in mind, or to know why this workaround was needed? Like a comment in the `tsup.config`? 🤔

File: api/tsup.config.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+import { defineConfig } from 'tsup';
+
+export default defineConfig({
+	entry: ['src/index.ts', 'src/cli/run.ts'],
+	format: 'esm',
+	async onSuccess() {
+		await cpy(['src/**/*.{yaml,liquid}'], 'dist');
+	},
+	dts: 'src/index.ts',
+});
```
Comment (jaads) [inline api/tsup.config.js:11]: this is effectively the workaround you mentioned right? Shall we add a comment as a reminder why this is needed with a reference to the issue  https://github.com/evanw/esbuild/pull/2508 ? 🤔

Comment (paescuj) [reply api/tsup.config.js:11]: No, this one is just a 1-to-1 replacement of the previous build script:
```json
	"build": "tsc --project tsconfig.prod.json && copyfiles \"src/**/*.{yaml,liquid}\" -u 1 dist",
```

The actual workarounds are applied at multiple places:

- **Workaround 1** - Context-depended paths for getting `yaml` / `liquid` files:
  - [`api/src/services/mail/index.ts`](https://github.com/directus/directus/pull/19089/files#diff-b65b152ce847c4566f72298678c2903bcb0f123ca8de71abb5743bd3e0be7b4d)
  - [`api/src/cli/utils/create-db-connection.ts`](https://github.com/directus/directus/pull/19089/files#diff-b65bd1e3d47d761595d4de9024f0df5773d9907ff99f68f76c20def71e47459b)
  - [`api/src/cli/utils/create-env/index.ts`](https://github.com/directus/directus/pull/19089/files#diff-8e6b4acbcb5a8b312a01b1c3d4bd156a5cff84dbc1ec862fac7554eb4f6bf8f8)
  - [`api/src/database/system-data/fields/index.ts`](https://github.com/directus/directus/pull/19089/files#diff-026ca6355fa8884f6e25bfc7f6cc8351ec67dfc33a5c674b8e8bf09793b00945)
  - [`api/src/database/system-data/collections/index.ts`](https://github.com/directus/directus/pull/19089/files#diff-ab42e6dd9097a8bb51184962cb4d5d08df30da719c7a41f0f2499d77baf935ca)
  - [`api/src/database/system-data/app-access-permissions/index.ts`](https://github.com/directus/directus/pull/19089/files#diff-5798bdac1e9a9acc4f624c0dbd8dbeb3eb18cae164b673108602577f7ae3b60d)
  - [`api/src/database/system-data/relations/index.ts`](https://github.com/directus/directus/pull/19089/files#diff-512f6b0837b355105727fa449a6abd84ff9b37139bccaa52b2d152c6ce758a45)
  - [`api/src/database/seeds/run.ts`](https://github.com/directus/directus/pull/19089/files#diff-ea5259e27174c13eb7e51ecdb72b21d57236f1fef9ddb95efb5ec7572707c833)
  All files are using `CONTEXT_ROOT` from [api/src/constants.ts](https://github.com/directus/directus/pull/19089/files#diff-5da24c390f8befd890a3c9e96c17862e386efabee8937b09120547bfcaa397d0) which is where the path is computed once using [findContextRoot](https://github.com/directus/directus/pull/19089/files#diff-e47968a1a045eda0ed8db3525151bb8a5cb7962067c87d8dba2f355af7e24fb4). So maybe we could add comment in this file?

- **Workaround 2** - Statically imported scripts:
  - Direct import of internal migrations ([`9875f1e`](https://github.com/directus/directus/pull/19089/commits/9875f1ebb62a9dac4c2d8c564785adeda69a0f89))
  -  Direct import of internal operations ([`b3bc237`](https://github.com/directus/directus/pull/19089/commits/b3bc237b1690fc72c9ac35282db3d085c6226861))
  Would you like to have a comment in both places?

Comment (jaads) [reply api/tsup.config.js:11]: I see 👍 👍 well, or better just a linear issue to keep track of it?

Comment (paescuj) [reply api/tsup.config.js:11]: Good idea! Added to backlog 👍

**Review by jaads — APPROVED** [review]
Looks all good and reasonable to me! build and tests are running successfully and the app operates fine 👍

**Comment by paescuj** [comment]
Will be split into several PRs and therefore closed.

**Pull request closed without merging**